SYCL/ESIMD/esimd_test_utils.hpp

```diff
@@ -99,26 +100,73 @@ bool write_binary_file(const char *fname, const std::vector<T> &vec,
 }
 
 template <typename T>
-bool cmp_binary_files(const char *fname1, const char *fname2, T tolerance) {
+bool cmp_binary_files(const char *fname1, const char *fname2, const T tolerance,
+                      const double toleratedMismatchRate = 0,
+                      const int mismatchReportThrottleLimit = 5) {
+
+  if (toleratedMismatchRate) {
+    if (toleratedMismatchRate >= 1 || toleratedMismatchRate < 0) {
+      std::cerr << "Tolerated mismatch rate (" << toleratedMismatchRate
+                << ") must be set within [0, 1) range" << std::endl;
+      return false;
+    }
+
+    std::cerr << "Tolerated mismatch rate set to " << toleratedMismatchRate
+              << std::endl;
+  }
+
   const auto vec1 = read_binary_file<T>(fname1);
   const auto vec2 = read_binary_file<T>(fname2);
+
   if (vec1.size() != vec2.size()) {
     std::cerr << fname1 << " size is " << vec1.size();
     std::cerr << " whereas " << fname2 << " size is " << vec2.size()
               << std::endl;
     return false;
   }
-  for (size_t i = 0; i < vec1.size(); i++) {
+
+  double totalMismatches = 0;
+  const double size = vec1.size();
+  for (size_t i = 0; i < size; i++) {
     if (abs(vec1[i] - vec2[i]) > tolerance) {
-      std::cerr << "Mismatch at " << i << ' ';
-      if (sizeof(T) == 1) {
-        std::cerr << (int)vec1[i] << " vs " << (int)vec2[i] << std::endl;
-      } else {
-        std::cerr << vec1[i] << " vs " << vec2[i] << std::endl;
+      if (!toleratedMismatchRate ||
+          (totalMismatches < mismatchReportThrottleLimit)) {
+
+        std::cerr << "Mismatch at " << i << ' ';
+        if (sizeof(T) == 1) {
+          std::cerr << (int)vec1[i] << " vs " << (int)vec2[i];
+        } else {
+          std::cerr << vec1[i] << " vs " << vec2[i];
+        }
+
+        if (!toleratedMismatchRate) {
+          std::cerr << std::endl;
+          return false;
+        } else {
+          std::cerr << ". Current mismatch rate: " << std::setprecision(8)
+                    << std::fixed << totalMismatches / size << std::endl;
+        }
+
+      } else if (totalMismatches == mismatchReportThrottleLimit) {
+        std::cerr << "Mismatch output throttled ... " << std::endl;
       }
+
+      totalMismatches++;
+    }
+  }
+
+  if (totalMismatches) {
+    const auto totalMismatchRate = totalMismatches / size;
+    if (totalMismatchRate > toleratedMismatchRate) {
+      std::cerr << "Mismatch rate of " << totalMismatchRate
+                << " has exceeded the tolerated amount of "
+                << toleratedMismatchRate << std::endl;
       return false;
     }
+
+    std::cerr << "Total mismatch rate is " << totalMismatchRate << std::endl;
```

Comment:
With introduction of tolerated mismatched rate, we need to also introduce and report maximum detected relative mismatch (in %). I would say if this is more than 10-20% in any element, that would indicate a problem, and test test developer should pay attention.

Reply:
>>  if this is more than 10-20% in any element, that would indicate a problem

That's subtle: interpretation depends on generator semantics, e.g. in case of overflow allowed during generation
for uint8_t what's the distance between 255 and 0? It may be considered as 255 in one direction, but also as 1+N*(255+1) if we go the other way over the circle, where N is from 0 to +inf. 

Hence, ambiguity.

Reply:
No, I mean relative, not absolute, just as I wrote. abs((expected-val)/expected)*100.
So, please report this info rather than silently returning "pass".

Note that I did not propose return "fail", irrespective of the max relative mismatch. I would like test developers or people who run tests see it and decide themselves if it is OK or not.

Reply:
> No, I mean relative, ... abs((expected-val)/expected)*100

Konstantin, I understand, but formally it's derived from absolute difference [ abs((expected-val) ...  ]
and it doesn't matter if later it's expressed as a percentage from the expected value or anything else: it gives **no practically useful information**.

Example. Let's look at some full error log:
...
byte 1: got 1 instead of 255
byte 2: got 10 instead of  9
byte 3: got 5 instead of 4
byte 100: got 12 instead of 9
byte 101: got 11 instead of  8
byte 102: got 6 instead of 3
...

With this info we can do at least something: observe the whole picture, try to infer a pattern.
But if we report max relative divergency for a single byte (byte 1 in this case) it will give no useful information by itself.

There's also one more thing: all of this is not directly related to the topic of this ticket.

Reply:
Konstantin,
It's not applicable in case of toleratedMismatchRate == 0 as we fail right away there.
In case of toleratedMismatchRate  != 0 log contains plenty of records that allow a user to make conclusions about deviations or increase print limit to observe the full log.

There's practically no use case for the suggested data.

I could understand adding something if there's a use case, but, I'm sorry, I can't understand adding more code if there's no foreseeable use case (taking into account all mentioned above).

Reply:
I outlined the use case above - please take a look.
Having user scan through all tolerated mismatches in case there are many of these is error prone.

Reply:
NB: in current implementation arguments fname1, fname2 are not assigned to any particular role of being reference or not.
Adding the requested printout will imply assigning these roles.